Source/playerdat.cpp

```diff
@@ -7,68 +7,272 @@
-	47622,
-	63364,
-	83419,
-	108879,
-	141086,
-	181683,
-	231075,
-	313656,
-	424067,
-	571190,
-	766569,
-	1025154,
-	1366227,
-	1814568,
-	2401895,
-	3168651,
-	4166200,
-	5459523,
-	7130496,
-	9281874,
-	12042092,
-	15571031,
-	20066900,
-	25774405,
-	32994399,
-	42095202,
-	53525811,
-	67831218,
-	85670061,
-	107834823,
-	135274799,
-	169122009,
-	210720231,
-	261657253,
-	323800420,
-	399335440,
-	490808349,
-	601170414,
-	733825617,
-	892680222,
-	1082908612,
-	1310707109
+namespace {
+
+struct ExperienceData {
+	/** Specifies the experience point limit of each level. The given values are defaults used if the data file is missing. */
+	std::vector<uint32_t> levelThresholds {
+		0,
+		2000,
+		4620,
+		8040,
+		12489,
+		18258,
+		25712,
+		35309,
+		47622,
+		63364,
+		83419,
+		108879,
+		141086,
+		181683,
+		231075,
+		313656,
+		424067,
+		571190,
+		766569,
+		1025154,
+		1366227,
+		1814568,
+		2401895,
+		3168651,
+		4166200,
+		5459523,
+		7130496,
+		9281874,
+		12042092,
+		15571031,
+		20066900,
+		25774405,
+		32994399,
+		42095202,
+		53525811,
+		67831218,
+		85670061,
+		107834823,
+		135274799,
+		169122009,
+		210720231,
+		261657253,
+		323800420,
+		399335440,
+		490808349,
+		601170414,
+		733825617,
+		892680222,
+		1082908612,
+		1310707109,
+		1583495809
+	};
+
+	static constexpr uint8_t DefaultMaxLevel = 50;
+	uint8_t maxLevel = DefaultMaxLevel;
+
+	[[nodiscard]] uint32_t getThresholdForLevel(unsigned level)
+	{
+		return levelThresholds[std::min<size_t>({ level, maxLevel, levelThresholds.size() - 1 })];
+	}
+} ExperienceData;
+
+struct ExperienceColumnDefinition {
+	enum class ColumnType {
+		Level,
+		Experience,
+		Unknown = -1
+	} type;
+
+	// The number of fields between this column and the last one identified as important (or from start of the record if this is the first column we care about)
+	unsigned skipLength;
+
+	static ColumnType mapNameToType(string_view name)
+	{
+		if (name == "Level") {
+			return ColumnType::Level;
+		}
+		if (name == "Experience") {
+			return ColumnType::Experience;
+		}
+		return ColumnType::Unknown;
+	}
+
+	ExperienceColumnDefinition() = delete;
+
+	ExperienceColumnDefinition(const ColumnType &type)
+	    : type(type)
+	    , skipLength(0)
+	{
+	}
+
+	ExperienceColumnDefinition(const ColumnType &type, unsigned skipLength)
+	    : type(type)
+	    , skipLength(skipLength)
+	{
+	}
+
+	bool operator==(const ExperienceColumnDefinition &other) const
+	{
+		return type == other.type && skipLength == other.skipLength;
+	}
 };
 
+void ReloadExperienceData()
+{
+	LoadDataResult dataFile = LoadDataFile("txtdata\\Experience.txt");
+	if (dataFile.hasError()) {
+		// Can't load the data file, so reset to defaults.
+		// TODO: Log error somewhere the player can see it, they probably want to know that we couldn't load the file
+		ExperienceData = {};
+		return;
+	}
+
+	constexpr unsigned ExpectedColumnCount = 2; // Level, Experience
+
+	std::vector<ExperienceColumnDefinition> columns;

```

Comment:
`std::array`?

Reply:
I'm gonna do this as a separate commit along with the std::unordered_set -> std::bitset change. Having to default-initialise all members of the array makes it very messy.

Reply:
You can use a `StaticVector` if you want to avoid the default initialization.